nifi-nar-bundles/nifi-media-bundle/nifi-media-processors/src/main/java/org/apache/nifi/processors/document/ExtractDocumentText.java

```diff
@@ -0,0 +1,92 @@
+import org.apache.nifi.processor.AbstractProcessor;
+import org.apache.nifi.processor.ProcessContext;
+import org.apache.nifi.processor.ProcessSession;
+import org.apache.nifi.processor.Relationship;
+import org.apache.nifi.processor.exception.ProcessException;
+import org.apache.tika.Tika;
+
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.OutputStreamWriter;
+import java.io.Reader;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+@Tags({"extract, text, pdf, word, excel, powerpoint, office"})
+@CapabilityDescription("Run Apache Tika text extraction to extra the text from supported binary file formats such as PDF " +
+        "and Microsoft Office files.")
+public class ExtractDocumentText extends AbstractProcessor {
+    private static final String TEXT_PLAIN = "text/plain";
+
+    public static final Relationship REL_ORIGINAL = new Relationship.Builder().name("original")
+            .description("On successful extraction, the original flowfile goes here").build();
+
+    public static final Relationship REL_EXTRACTED = new Relationship.Builder().name("extracted")
+            .description("On successful extraction, the extracted content goes here").build();
+
+    public static final Relationship REL_FAILURE = new Relationship.Builder().name("failure")
+            .description("Content extraction failed").build();
+
+    private Set<Relationship> relationships = Collections.unmodifiableSet(new HashSet<>(Arrays.asList(REL_ORIGINAL, REL_EXTRACTED, REL_FAILURE)));
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return this.relationships;
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        FlowFile flowFile = session.get();
+        if (flowFile == null) {
+            return;
+        }
+
+        FlowFile extracted = session.create(flowFile);
+        boolean error = false;
+        try (InputStream is = session.read(flowFile);
+             Reader tikaReader = new Tika().parse(is);
+             OutputStream os = session.write(extracted);
+             OutputStreamWriter writer = new OutputStreamWriter(os)) {
+            IOUtils.copy(tikaReader, writer);
+        } catch (final Throwable t) {
+            error = true;
+            getLogger().error("Extraction Failed {}", flowFile, t);
+            session.remove(extracted);
+            session.transfer(flowFile, REL_FAILURE);
+        } finally {
+            if (!error) {
+                Map<String, String> mimeAttrs = new HashMap<>();
+                mimeAttrs.put("mime.type", TEXT_PLAIN);
+                extracted = session.putAllAttributes(extracted, mimeAttrs);
+                session.transfer(extracted, REL_EXTRACTED);
+                session.transfer(flowFile, REL_ORIGINAL);
+            }
```

Comment:
@MikeThomsen Is there a reason for setting and checking the `error` flag versus just calling the successful transfer in the try block? Otherwise this looks good to go.

```suggestion
        try (InputStream is = session.read(flowFile);
             Reader tikaReader = new Tika().parse(is);
             OutputStream os = session.write(extracted);
             OutputStreamWriter writer = new OutputStreamWriter(os) ) {
            IOUtils.copy(tikaReader, writer);
            Map<String, String> mimeAttrs = new HashMap<>();
            mimeAttrs.put("mime.type", TEXT_PLAIN);
            extracted = session.putAllAttributes(extracted, mimeAttrs);
            session.transfer(extracted, REL_EXTRACTED);
            session.transfer(flowFile, REL_ORIGINAL);
        } catch (final Throwable t) {
            getLogger().error("Extraction Failed {}", flowFile, t);
            session.remove(extracted);
            session.transfer(flowFile, REL_FAILURE);
        }
```

Reply:
> Is there a reason for setting and checking the error flag versus just calling the successful transfer in the try block?

I don't know if it was fixed recently, but traditionally, `transfer` wouldn't work if the input/output streams are still open. Didn't want to defeat the purpose of try-with block.

Reply:
Thanks for the clarification.